### PR TITLE
[python] Reject lone surrogates instead of crashing the AST emitter

### LINKED ISSUE #6740: [python] Lone surrogate in a string literal crashes the AST emitter
## Summary

A string literal containing a lone surrogate crashes the AST generator with an uncaught `UnicodeEncodeError` traceback. Lone surrogates are valid in CPython `str` objects, and the CPython test suite uses them extensively to exercise codec edge cases.

## Reproducer

```python
s = "\ud800"
assert len(s) == 1
```

## Observed

```
  File ".../parser/json_emitter.py", line 168, in generate_ast_json
    json.dump(ast_json, json_file, indent=4, ensure_ascii=False)
  File "/usr/lib/python3.12/json/__init__.py", line 180, in dump
    fp.write(chunk)
UnicodeEncodeError: 'utf-8' codec can't encode character '\ud800' in position 1: surrogates not allowed
```

Exit code `1`, full Python traceback on stderr. ESBMC 8.4.0.

CPython 3.12.3 runs the file to completion, exit 0.

## Cause

`json_emitter.generate_ast_json` writes the AST with `ensure_ascii=False`. A lone surrogate is representable in a Python `str` but not encodable as UTF-8, so `json.dump` raises. Using `ensure_ascii=True` (the `json` default, which escapes non-ASCII as `\uXXXX`) or opening the file with `errors="surrogatepass"` would both round-trip these literals.

## Category

**ESBMC defect.** The frontend crashes on valid Python input; it does not merely decline to verify it.

## Affected CPython tests (18)

`test_bigmem`, `test_cgi`, `test_cmd_line`, `test_codeccallbacks`, `test_embed`, `test_float`, `test_genericpath`, `test_int`, `test_memoryio`, `test_multibytecodec`, `test_pathlib`, `test_plistlib`, `test_posixpath`, `test_pyexpat`, `test_stringprep`, `test_tarfile`, `test_tcl`, `test_userstring`

### PR DESCRIPTION
**Pull request by rafaelsamenezes**
A string literal containing a lone surrogate crashed the parser with an
uncaught `UnicodeEncodeError` traceback. Neither transportable spelling
survives the C++ reader, so the construct is refused at emission with a
located diagnostic (exit 4) instead of aborting mid-`json.dump`.

Full surrogate support would need a cross-boundary encoding for strings
that have no UTF-8 form; that is an enhancement, not part of this fix.

Fixes #6740

### CODE REVIEW

File: src/python-frontend/parser/json_emitter.py [hunk 1]
```diff
@@ -141,6 +142,28 @@ def _compute_output_json_path(
     return os.path.join(output_dir, f"{os.path.basename(python_filename[:-3])}.json")
 
 
+def _reject_unencodable_string_constants(tree: ast.AST, source_filename: str) -> None:
```
[the remark this answers is not in the document]
Comment (lucasccordeiro) [reply src/python-frontend/parser/json_emitter.py:145]: Keeping `print` + `sys.exit(4)` deliberately — it is the established parser-stage convention here, not an ad-hoc choice. `threading_lowering._emit_error` is byte-for-byte this shape (`print(f"ERROR: {file}:{line}: {msg}")` then `sys.exit(4)`), and `import_resolver` prints `ERROR: ...` to stdout as well. `python_language.cpp` spawns the parser with `bp::child` and no stream redirection, then does `exit(process.exit_code())`, so exit 4 is the wire protocol for "unsupported construct at parse time" and the diagnostic reaches the user either way.

Raising and unwinding to `parser_cli` would need the same exit code re-derived at the top level and would leave this one helper inconsistent with its two siblings. On testability: neither `json_emitter` nor `threading_lowering` has unit tests today (`unit/python-frontend/` covers the preprocessor and the models), so there is nothing this blocks right now. If we want the exception-based shape it is worth doing across all three call sites in its own change rather than diverging here.

Applied the other two comments in b830963 — the diagnostic now names the offending code point and carries the column.

File: src/python-frontend/parser/json_emitter.py [hunk 2]
```diff
@@ -141,6 +142,28 @@ def _compute_output_json_path(
...
+    """Refuse string literals that cannot cross the parser->converter JSON boundary.
+
+    A lone surrogate is representable in a CPython ``str`` but has no UTF-8
+    encoding, and the C++ reader rejects both transportable spellings: an
+    escaped ``\\ud800`` is invalid JSON (nlohmann requires a low surrogate to
+    follow), and the WTF-8 byte form is ill-formed UTF-8. Emitting one aborts
```
[the remark this answers is not in the document]
Comment (lucasccordeiro) [reply src/python-frontend/parser/json_emitter.py]: Correct on the wording, and fixed in b830963.

The detection was already right — Python strings are sequences of code points and CPython never combines a surrogate pair, so `"😀"` is two unencodable code points rather than an emoji, and rejecting it matches what `json.dump` would have crashed on. Only the message was misleading. It now reports what was actually found:

```
ERROR: main.py:1:4: string literal contains surrogate code point U+D83D, which has
no UTF-8 encoding and is not supported by ESBMC's Python frontend
```

The code point comes from `UnicodeEncodeError.start`, so it names the first offending character rather than a generic category. Added `github_6740_surrogate_pair_unsupported` to pin the pair case alongside the lone-surrogate one, and the docstring now states the no-pairing rule explicitly.

File: src/python-frontend/parser/json_emitter.py [hunk 3]
```diff
@@ -141,6 +142,28 @@ def _compute_output_json_path(
...
+    the parser mid-``json.dump``, so refuse at parse time with a located
+    diagnostic rather than crash or emit a silently truncated string.
+    """
+    for node in ast.walk(tree):
+        if not isinstance(node, ast.Constant) or not isinstance(node.value, str):
+            continue
+        try:
+            node.value.encode("utf-8")
+        except UnicodeEncodeError:
+            print(f"ERROR: {source_filename}:{getattr(node, 'lineno', '?')}: "
```
[the remark this answers is not in the document]
Comment (lucasccordeiro) [reply src/python-frontend/parser/json_emitter.py]: Applied in b830963 — the diagnostic is now `file:line:col`.

Passing `col_offset` through unchanged (0-based) rather than converting to 1-based, to match how the frontend already reports Python locations: `converter_util.cpp:71` does `location.set_column(ast_node["col_offset"].get<int>())` verbatim, so a column quoted by the parser and one quoted by the converter now agree. Left `end_col_offset` out — the start is enough to disambiguate literals on a shared line, and the span would not narrow it further.

File: src/python-frontend/parser/json_emitter.py [hunk 4]
```diff
@@ -141,6 +142,28 @@ def _compute_output_json_path(
...
+                  "string literal contains a lone surrogate, which is not supported by "
+                  "ESBMC's Python frontend")
+            sys.exit(4)
```
[the remark this answers is not in the document]
Comment (lucasccordeiro) [reply src/python-frontend/parser/json_emitter.py]: Same point as the thread above — answered in https://github.com/esbmc/esbmc/pull/6747#discussion_r3722341936.

File: src/python-frontend/parser/json_emitter.py [hunk 5]
```diff
@@ -141,6 +142,28 @@ def _compute_output_json_path(
...
+    diagnostic rather than crash or emit a silently truncated string.
+    """
+    for node in ast.walk(tree):
+        if not isinstance(node, ast.Constant) or not isinstance(node.value, str):
+            continue
+        try:
+            node.value.encode("utf-8")
+        except UnicodeEncodeError:
+            print(f"ERROR: {source_filename}:{getattr(node, 'lineno', '?')}: "
+                  "string literal contains a lone surrogate, which is not supported by "
```
[the remark this answers is not in the document]
Comment (lucasccordeiro) [reply src/python-frontend/parser/json_emitter.py]: Same point as the thread above — answered in https://github.com/esbmc/esbmc/pull/6747#discussion_r3722343170.
